Use cacheKeyForTree to improve the addon dedup

dnachev opened the pull request:
Attempt to fix #416 .

The change keeps the old logic for all invocations, which are not `treeFor`. For `treeFor`, the code will consult `cacheKeyForTree` to determine whether the trees from the host addon and the engine dependency addon are the same.

There is an open question of whether finding the corresponding host addon simply based on the name is enough. The host might have multiple instances of the same addon and we will only check for duplication against the first encountered one. Previously, the name was all cared about, so it worked just fine.

trentmwillis commented on lib/engine-addon.js:
```diff
@@ -8,6 +8,7 @@ var writeFile = require('broccoli-file-creator');
 var Babel = require('broccoli-babel-transpiler');
 var concat = require('broccoli-concat');
 var DependencyFunnel = require('broccoli-dependency-funnel');
+var defaults = require('lodash.defaults');
```
Can we just use `Object.assign` instead? I'm fairly certain we only need to support Node 4+ at this point due to our Ember-CLI support policy

rwjblue replied:
we can, but we'd have to bump to 0.6.0 (since 0.5.0 supported 0.12 when it released)

trentmwillis replied:
But Ember-CLI 2.12 doesn't support Node 4, does it? I don't particularly care, I was just under the impression that our support policy was based on the Ember(-CLI) versions we supported. Regardless, now seems as good a time as any to drop 0.12.

rwjblue replied:
Agreed all around. I had forgotten that we required Ember-cli@2.12 with ember-engines@0.5.0. Seems like 0.5 dropped node 0.12 support for us. 🎊

trentmwillis commented on node-tests/acceptance/build-test.js:
```diff
@@ -398,5 +398,61 @@ describe('Acceptance', function() {
       output.doesNotContain(`engines-dist/${engineName}/assets/engine-vendor.css`, cssCommentMatcher(`${addonName}.css`));
     }));
+
+    it('duplicate addons in lazy engines when tree is not the same', co.wrap(function* () {
```
This would be clearer as: `duplicates an addon in lazy engines when the tree returned by the addon is different`

trentmwillis commented on lib/engine-addon.js:
```diff
@@ -358,12 +356,33 @@ module.exports = {
+                return;
+              }
+              break;
+            case 'included':
```
Why is `included` special cased here?

dnachev replied:
Just an explicit list of the cases, but I guess for only two cases makes for rather unordinary looking switch statement.

trentmwillis commented on lib/engine-addon.js:
```diff
@@ -281,6 +282,15 @@ var buildEngineStyleTree = memoize(function buildEngineStyleTree() {
+ * addons are deduped. Different instances of the same addon will
+ * be detected based on this identifier.
+ */
+function generateAddonId(addon) {
```
Why do we need to call this `addonId` instead of just using `addon.name`? The function isn't doing anything special, and by introducing a new concept for identifying addons, it makes the code much harder to reason about.

dnachev replied:
The initial idea was to encapsulate a better algorithm to determine the duplicate addon, which would be based on something more than just the name. However, considering that ember-cli doesn't have any notion of addon versioning in the same project, I think it was unnecessary. I will switch back to just use the name of the addon.

dnachev commented:
@trentmwillis I think I've addressed your comments.

I will try it out on our application to further proof it works once we have agreed this is the correct approach.

dnachev commented:
It seems our application still works with this fix. Is there anything else holding this out?

dnachev commented:
I would love another look at the changes to see whether there is anything more to be improved? @rwjblue @trentmwillis
